rust/kernel/file_operations.rs

```diff
@@ -249,6 +306,88 @@ macro_rules! declare_file_operations {
     };
 }
 
+/// Allows the handling of ioctls defined with the `_IO`, `_IOR`, `_IOW`, and `_IOWR` macros.
+///
+/// For each macro, there is a handler function that takes the appropriate types as arguments.
+pub trait IoctlHandler: Sync {
+    /// Handles ioctls defined with the `_IO` macro, that is, with no buffer as argument.
+    fn pure(&self, _file: &File, _cmd: u32, _arg: usize) -> KernelResult<i32> {
+        Err(Error::EINVAL)
+    }
+
+    /// Handles ioctls defined with the `_IOR` macro, that is, with an output buffer provided as
+    /// argument.
+    fn read(&self, _file: &File, _cmd: u32, _writer: &mut UserSlicePtrWriter) -> KernelResult<i32> {
+        Err(Error::EINVAL)
+    }
+
+    /// Handles ioctls defined with the `_IOW` macro, that is, with an input buffer provided as
+    /// argument.
+    fn write(
+        &self,
+        _file: &File,
+        _cmd: u32,
+        _reader: &mut UserSlicePtrReader,
+    ) -> KernelResult<i32> {
+        Err(Error::EINVAL)
+    }
+
+    /// Handles ioctls defined with the `_IOWR` macro, that is, with a buffer for both input and
+    /// output provided as argument.
+    fn read_write(&self, _file: &File, _cmd: u32, _data: UserSlicePtr) -> KernelResult<i32> {
+        Err(Error::EINVAL)
+    }
+}
+
+/// Represents an ioctl command.
+///
+/// It can use the components of an ioctl command to dispatch ioctls using
+/// [`IoctlCommand::dispatch`].
+pub struct IoctlCommand(u32);
+
+impl IoctlCommand {
+    /// Dispatches the given ioctl to the appropriate handler based on the value of the command. It
+    /// also creates a [`UserSlicePtr`], [`UserSlicePtrReader`], or [`UserSlicePtrWriter`]
+    /// depending on the direction of the buffer of the command.
+    ///
+    /// It is meant to be used in implementations of [`FileOperations::ioctl`] and
+    /// [`FileOperations::compat_ioctl`].
+    ///
+    /// # Safety
+    ///
+    /// The caller must ensure that `fs` is set to a value compatible with the pointer in `arg`.
+    /// This is because this function uses [`UserSlicePtr`], which calls `access_ok`.
+    pub unsafe fn dispatch<T: IoctlHandler>(
+        &self,
+        handler: &T,
+        file: &File,
+        arg: usize,
```

Comment:
If `arg` was a parameter to the `IoctlCommand` constructor, would we be able to make this function safe, because the caller wouldn't be able to use provide an incorrect value?

Reply:
I actually like the idea of moving `arg` into `IoctlCommand`, but it won't allow us to make this function safe because one could still call `set_fs`, then call `dispatch` and end up with `access_ok` not checking a buffer that it should.

Oh, we could also capture the expected `fs` in `IoctlCommand` during construction and ensure we're with the right one at dispatch time. That'll make ioctls slightly more expensive but I think worth it for the extra guarantee.

Reply:
If we did the `access_ok` in construction, but only returned an error if you tried to actually _use_ it as a pointer, that should also work, right, since the kernel will only invoke `compat_ioctl` and `unlocked_ioctl` in the right `fs`.

Reply:
Yes. I didn't want to do that because we don't know if `cmd` is a standardised (for the driver using it), so the pointer and size we get from it are not necessarily meaningful. (We know they are by the time we call `dispatch` because it is only called in such cases.)

But now I realise we can call `access_ok`, cache the result, then acting on it only if `dispatch` is called.

Reply:
I made this change, `dispatch` is now safe. PTAL.